### DISCUSSION TITLE
Enabling sticky sessions with nginx ingress in front of keycloak on kubernetes

### DISCUSSION THREAD

**Discussion by cjohn001** [keycloak/keycloak#23615] [Archive]
Hello together,
I am currently trying to figure out how I can set sticky sessions in my nginx-ingress running in front of keycloakx. I have keycloak running now in proxy mode reencrypt. Enabling sticky sessions is the last thing which is missing. However, unfortunately I have not found any examples on how to do it and the documentation is not clear to me.
I tried the following as I understood that nginx needs to decide based on the AUTH_SESSION_ID cookie how to route requests. 
Unfortunately,  with this configuration nginx-ingress sets a new cookie, rather than using the values in the cookie provided by keycloak. Would be great if someone could provide the required config options. Thank you very much!

My try:

```
nginx.ingress.kubernetes.io/affinity: 'cookie'
nginx.ingress.kubernetes.io/session-cookie-name: 'AUTH_SESSION_ID'
nginx.ingress.kubernetes.io/session-cookie-expires: '172800'
nginx.ingress.kubernetes.io/session-cookie-max-age: '172800'
```

**Comment by shawkins** [comment]
@cjohn001 in looking over the nginx docs and the corresponding feature exposed by openshift routes it appears that cookie is managed at the proxy level.

For nginx to manage AUTH_SESSION_ID in the way you are wanting, it would need to maintain a table of those id's to the consistent hash of the upstream used for the request.  That is likely something that won't be supported.

**Comment by cjohn001** [comment]
Hello @shawkins,
thanks for your reply. So if I understand you correctly, you say sticky sessions are not supported with keyclaok in combination with nginx? This is indeed a pitty :( 

I am wondering, which load balancer is used in openshift and how does this have to be configured? I assume I could also switch to a different load balancer or even an istio service mesh in my kubernetes cluster. Unfortunately, the keycloak documentation is rather insufficient in regards to sticky session config details.

Another question in this regards: If I understood the keycloak documentation correctly, when opening a session, keycloak will decide which keycloak instance shall process further requests for this session and writes the relevant instance id to the AUTH_SESSION_ID cookie. So lets assume I would let nginx attach a cookie to the session as well. Than it will likely route to a different instance than keycloak selected. However, nginx will than always route to the same instance. Would this than denote, that  the receiving keycloak instance would each time have to look up session details from the infinispan cache? Or would it only do this the first time? 
In case it than keeps the session data, I assume I would only have one cash miss the very first time after session creation. So performance wise not a great issue when using a standard nginx sticky session cookie?

**Reply by shawkins** [reply]
> So if I understand you correctly, you say sticky sessions are not supported with keyclaok in combination with nginx?

It might be good to review #8692 and #16893 - in particular what was added to the guide https://github.com/keycloak/keycloak/blob/main/docs/guides/server/reverseproxy.adoc#enable-sticky-sessions

"You need to configure your loadbalancer to sticky over the AUTH_SESSION_ID cookie. How exactly do this is dependent on your loadbalancer.

If your proxy supports session affinity without processing cookies from backend nodes, you should set the spi-sticky-session-encoder-infinispan-should-attach-route option to false in order to avoid attaching the node to cookies and just rely on the reverse proxy capabilities."

Seems to imply that if the load balancer cannot use a backend cookie, you should just let the load balancer handle things on its own.

**Reply by cjohn001** [reply]
thanks for the directions
